### DISCUSSION TITLE
Using at_object_creation after prototype creation

### DISCUSSION THREAD

**Discussion by RodRitter** [evennia/evennia#2700] [Evennia Questions] [answered]
Good day all, it seems I've been posting a lot here lately so I hope another question isn't too much.

I've got a `NPC(Character)` class that updates the `hp` value of an NPC depending based on rules defined by other attribtues. The problem is that if I set these attributes using the prototype system, it doesn't apply my `hp` rule afterward. Only before. See below

```
RAT = {
   [...]
    "strength": 5,
    "agility": 5,
    "spirit": 5,
    "constitution": 10,
}
```
```
class Character(DefaultCharacter):
    def at_object_creation(self):
        self.db.hp = 5 # <-- default health
        ...

class NPC(Character):
    def at_object_creation(self):
        super().at_object_creation()
        self.db.hp = rules.get_max_hp(self) # Should be CON * 5
        ...
```

So if I spawn a RAT, it will have an HP of 5, but it should be updating it according to its Constitution, which shows that the `at_object_creation` is happening either before the attributes are applied, or not at all. I need the hp to be applied after the stats are generated by the prototype spawner.

Any help is appreciated

**Comment by aMiss-aWry** [comment]
I have a similar issue, actually! I did this by adding a delayed call to set_description (for me, it was updating a description after the object was created).

from evennia.utils.utils import delay

Then in at_object_creation something like:
delay(1, self.set_description)

Then write a function in the NPC class like

def set_description(self):
(Whatever code here that relies on certain attributes to already be set)

For you instead of 'description' it could be set_stats or so on. What this does is a second after making the object, it updates the object's stats.

**Reply by RodRitter** [reply]
Thanks for your answer. While this solution works, I think @Griatch had a good point about setting it explicitly after creation.

**Comment by Griatch** [comment, marked as answer]
This is not how prototypes work - the prototype values always apply and override what you have in the hooks (that is, after the hook has run). 

It's equivalent to doing `create_object("Character", attributes=("hp", 10))` - you will always get the value you apply explicitly like that over what you set in the hook.

You need to apply the change manually after the prototype (no need for `delay` like the other answer, that's an unstable solution). 

```python
rat  = prototypes.spawn("Rat")
rat.start_setup()  # some setup hook of your own design 
```

You can of course also set the hp etc manually. The point is that you need to make your setup after the prototype creation. 

We don't have a method explicitly called 'after' a prototype spawn, maybe that would be an interesting thing to add. :thinking:

**Reply by RodRitter** [reply]
Thank you for your help. I think I'll add it as a lambda in the prototypes.py file then.

**Reply by RodRitter** [reply]
I think it'll definitely be helpful the have an "after_creation" type hook, especially if builders are manually adding objects via the command line
